sale_project_user_profile/models/sale_order.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api, _
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    @api.constrains(
+        'state',
+        'order_line',
+    )
+    def _check_state(self):
+        if self.state == 'sale':
+            service_products = self.mapped(
+                'order_line.product_id'
+            ).filtered(
+                lambda p: p.type == 'service'
+            )
+            track_services = service_products.mapped('track_service')
+            if 'user_profile' in track_services:
+                if any(s != 'user_profile' for s in track_services):
+                    raise ValidationError(_(
+                        "With a service product of type 'User profile', "
+                        "all service products must have 'User profile' type."
+                    ))
+
+    @api.multi
+    def action_confirm(self):
+        # Copy of the same implementation in sale_timesheet module,
+        # but with track_service = 'user_profile'
+        result = super(SaleOrder, self).action_confirm()
+        for order in self:
```

Comment:
Maybe you could reduce nesting with something of the sort and avoid the loop break also:
```python
for order in self:
    if order.project_project_id:
        continue
    if order.order_line.filtered(
        lambda x: x.track_service == 'user_profile')[0]:
```

Reply:
Use of filtered is less efficient because for break with first match